In [2]:
import pandas as pd
import numpy as np
import altair as alt

In [3]:
path = r'/Users/pravinkumar/Downloads/spotify_artists.csv'
df = pd.read_csv(path)

In [4]:
df.shape

(69414, 15)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69414 entries, 0 to 69413
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   artist_id           69414 non-null  object 
 1   name                69414 non-null  object 
 2   name_normalized     69104 non-null  object 
 3   country_of_origin   0 non-null      float64
 4   career_start_year   69414 non-null  int64  
 5   primary_genre_l1    69414 non-null  object 
 6   primary_genre_l2    69414 non-null  object 
 7   spotify_followers   1040 non-null   float64
 8   spotify_popularity  1040 non-null   float64
 9   monthly_listeners   0 non-null      float64
 10  musicbrainz_id      0 non-null      float64
 11  lastfm_listeners    0 non-null      float64
 12  lastfm_play_count   0 non-null      float64
 13  is_solo_artist      0 non-null      float64
 14  source_dataset      69414 non-null  object 
dtypes: float64(8), int64(1), object(6)
memory usage: 7.9+

### Remvoving blank columns 

In [6]:
df = df.drop(columns =['country_of_origin', 'monthly_listeners', 'musicbrainz_id','lastfm_listeners','lastfm_play_count','is_solo_artist'])

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69414 entries, 0 to 69413
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   artist_id           69414 non-null  object 
 1   name                69414 non-null  object 
 2   name_normalized     69104 non-null  object 
 3   career_start_year   69414 non-null  int64  
 4   primary_genre_l1    69414 non-null  object 
 5   primary_genre_l2    69414 non-null  object 
 6   spotify_followers   1040 non-null   float64
 7   spotify_popularity  1040 non-null   float64
 8   source_dataset      69414 non-null  object 
dtypes: float64(2), int64(1), object(6)
memory usage: 4.8+ MB


In [8]:
df.index

RangeIndex(start=0, stop=69414, step=1)

In [9]:
df.describe().T


,count,mean,std,min,25%,50%,75%,max
career_start_year,69414.0,2.010581e+03,7.813138e+00,1957.0,2004.0,2011.0,2017.00,2023.0
spotify_followers,1040.0,6.695169e+06,1.392190e+07,1502.0,627495.0,2052725.5,7054540.25,145489371.0
spotify_popularity,1040.0,6.553558e+01,1.246812e+01,12.0,58.0,66.0,74.00,100.0


#### Q1: How many unique artists are present in the dataset?

In [10]:
df['name']= df['name'].fillna('Unknown')

In [11]:
unique_artists = len(df['name'].unique())
unique_artists

69414

### Findings : 
- There are 69414 artists present in the dataset.

#### Q2 : Which primary music genres (primary_genre_l1) have the highest number of artists?

In [12]:
genres_sorted = df['primary_genre_l1'].value_counts().reset_index(name='count')

In [13]:

genres_sorted

,primary_genre_l1,count
0,Electronic,15800
1,Rock,8644
2,Pop,6707
3,World,6237
4,Metal,5202
5,Latin,4410
6,R&B,3904
7,Folk,3508
8,Classical,2927
9,Hip-Hop,2381


In [14]:
alt.Chart(genres_sorted).mark_bar().encode(
    x = alt.X('primary_genre_l1:N', sort=alt.SortField(field = 'count', order= 'descending')),
    y= 'count:Q',
    tooltip = ['primary_genre_l1','count'],
    color = 'primary_genre_l1'
)

alt.Chart(...)

### Findings : 
- Electronic has the highest number of artists.

#### Q3 : Which secondary genres (primary_genre_l2) are the most common?

In [15]:
sub_genere = df['primary_genre_l2'].value_counts().head(10).reset_index(name = 'count')

In [16]:
sub_genere

,primary_genre_l2,count
0,Hip-Hop,2357
1,Jazz,2274
2,EDM,1849
3,Pop,1759
4,French,1716
5,Folk,1688
6,Gospel,1568
7,German,1530
8,Chill,1496
9,Funk,1434


In [17]:
alt.Chart(sub_genere).mark_bar().encode(
    x = 'count:Q',
    y = alt.Y('primary_genre_l2:N', sort= alt.SortField(field= 'count', order= 'ascending')),
    tooltip= ['primary_genre_l2', 'count'],
    color= 'primary_genre_l2'
)

alt.Chart(...)

#### Findings : 
- Hip-Hop is the most common secondary genere.

### Q4 : What percentage of all artists belongs to each primary genre?

In [18]:
artist_count = df['primary_genre_l1'].value_counts().reset_index(name = 'total_artists')


In [19]:
artist_count['artist_pct'] = artist_count['total_artists']/ df['name'].count()*100

In [20]:
artist_count.head()

,primary_genre_l1,total_artists,artist_pct
0,Electronic,15800,22.761979
1,Rock,8644,12.452819
2,Pop,6707,9.662316
3,World,6237,8.985219
4,Metal,5202,7.494165


In [21]:
artist_count['artist_pct'] = artist_count['artist_pct'].apply(lambda x : round(x, 2))

In [22]:
artist_count.head()

,primary_genre_l1,total_artists,artist_pct
0,Electronic,15800,22.76
1,Rock,8644,12.45
2,Pop,6707,9.66
3,World,6237,8.99
4,Metal,5202,7.49


In [23]:
alt.Chart(artist_count).mark_arc(innerRadius= 50).encode(
    theta='artist_pct:Q',
    color = 'primary_genre_l1:N',
    tooltip = ['primary_genre_l1','artist_pct']
)

alt.Chart(...)

#### Percentage of artists in each genre(Top 3) :
- Electronic = 22.76
- Rock = 12.45
- Pop = 9.66

### Q5 : How has the number of artists changed based on their career start year?

In [24]:
artists_by_year = df['career_start_year'].value_counts().sort_index().reset_index(name = 'no_of_artist').sort_values('no_of_artist', ascending = True)

In [25]:
artists_by_year

,career_start_year,no_of_artist
1,1958,1
3,1961,1
4,1962,1
0,1957,2
2,1960,3
...,...,...
43,2001,3560
59,2017,3616
60,2018,3812
61,2019,5375


In [26]:
alt.Chart(artists_by_year).mark_line(strokeWidth=7,color='orange').encode(
    x= 'career_start_year:N',
    y= 'no_of_artist:Q',
    tooltip = ['career_start_year','no_of_artist']
).properties(
    width = 600,
    height= 400
)

alt.Chart(...)

#### Findings : 
- No. of artists in each year were below 100 till 1999. 
- There was a huge surge in the number of artists (5535) in the year 2000.
- After that it dipped down to 2355 in the year 2003 and ramained almost flat till 2016 (2633).
- in 2019 we can see the rise in number of artists (5375).
- Since then it has been in the decreasing path.

#### Q6: Which primary genres have the oldest and newest average career start years?

In [27]:
avg_career_year_genre = df.groupby('primary_genre_l1')['career_start_year'].mean().round(0).sort_values()

In [28]:
avg_year_genre = avg_career_year_genre.reset_index(name= 'avg_year')
avg_year_genre

,primary_genre_l1,avg_year
0,Soundtrack,2005.0
1,Blues,2006.0
2,Country,2009.0
3,Rock,2009.0
4,Instrumental,2009.0
5,Unknown,2009.0
6,Classical,2010.0
7,Folk,2010.0
8,Gospel,2010.0
9,Reggae,2010.0


In [50]:
alt.Chart(avg_year_genre).mark_point(size=80, opacity=1.0).encode(
    x= 'primary_genre_l1:N',
    y= 'avg_year:N',
    color= 'primary_genre_l1:N'
).properties(
    width = 500,
    height = 400
)

alt.Chart(...)

### Q7 : Which years had the highest number of artists starting their careers?

In [30]:
highest_career_year = df['career_start_year'].value_counts().head(1)

In [31]:
highest_career_year

career_start_year
2000    5535
Name: count, dtype: int64

### Findings : 
- The highest number of artists(5535) started their career in the year 2000.

### Q8 : How has the popularity of different music genres changed across career-start periods?

In [32]:
# music_genre_popularity = df.groupby(['career_start_year','primary_genre_l1']).size().reset_index(name="artists")

In [33]:
music_genre_popularity = df.groupby(
    ['career_start_year','primary_genre_l1']
    )['primary_genre_l1'].size().reset_index(
        name="artists"
        ).sort_values(
            ['career_start_year','artists'], ascending = [True, False]
            ).drop_duplicates(
                'career_start_year'
                )

In [34]:
music_genre_popularity

,career_start_year,primary_genre_l1,artists
0,1957,Blues,1
2,1958,Rock,1
3,1960,Blues,2
5,1961,R&B,1
6,1962,R&B,1
...,...,...,...
629,2019,Electronic,1265
646,2020,Electronic,611
663,2021,Electronic,413
681,2022,Electronic,243


In [52]:
alt.Chart(music_genre_popularity).mark_bar().encode(
    x = 'career_start_year:N',
    y = 'primary_genre_l1:N',
    tooltip = ['career_start_year', 'primary_genre_l1','artists'],
    color = 'primary_genre_l1:N'
).properties(
    width =650,
    height = 400
)

alt.Chart(...)

### Findings :
- Since 1957-1989, 'Rock' has been the mostly popular genre
- From 1990-1996, 'R&B' was mostly popular.
- From 2000-2022, 'Electronic' has been the most popular genre.

#### Q9 : Which combinations of primary_genre_l1 and primary_genre_l2 occur most frequently?

In [36]:
freq_genre_combo = df.groupby(
    ['primary_genre_l1','primary_genre_l2']
    ).size().reset_index(name = 'occurence').sort_values(by= 'occurence', ascending = False).head(5)

In [37]:
freq_genre_combo

,primary_genre_l1,primary_genre_l2,occurence
69,Hip-Hop,Hip-Hop,2352
72,Jazz,Jazz,2261
26,Electronic,EDM,1849
107,Pop,Pop,1755
158,World,French,1695


In [38]:
freq_genre_combo['Combo_genre'] = freq_genre_combo['primary_genre_l1'] + '+' + freq_genre_combo['primary_genre_l2']

In [53]:
alt.Chart(freq_genre_combo).mark_bar().encode(
    y = alt.Y('Combo_genre:N', sort = alt.SortField('occurence', order = 'descending')),
    x = 'occurence:Q',
    color= 'Combo_genre:N',
    tooltip= ['primary_genre_l1','primary_genre_l2','occurence']
).properties(
    width = 500,
    height= 220
)

alt.Chart(...)

#### Based on the data :
- Primary Genre: Hip Hop & Secondary Genre: Hip Hop occurs most frequently.
- Primary Genre: Jazz & Secondary Genre: Jazz is at the 2nd position.
- Primary Genre: Electronic & Secondary Genre: EDM is at the 3rd position.

#### 10 : Which secondary genres are most diverse within each primary genre?

In [40]:
sec_genre_diversity = df.groupby(
    'primary_genre_l1'
    )['primary_genre_l2'].nunique().reset_index(
        name= 'sec_gen_count'
        ).sort_values(
            by='sec_gen_count', ascending= False
            )

In [41]:
sec_genre_diversity

,primary_genre_l1,sec_gen_count
3,Electronic,42
14,Rock,26
11,Pop,16
12,R&B,13
9,Latin,12
10,Metal,11
1,Classical,9
4,Folk,8
6,Hip-Hop,6
17,World,5


In [42]:
alt.Chart(sec_genre_diversity).mark_bar().encode(
    x = 'primary_genre_l1:N',
    y = 'sec_gen_count:Q',
    color = 'primary_genre_l1:N',
    tooltip = ['primary_genre_l1', 'sec_gen_count']
).properties(
    width = 600,
    height = 400
)

alt.Chart(...)

#### Findings : 
- Based on the findings, primary genre with most associated secondary genres is 'Electronic' with 42 different secondary genre associated with it.

#### Q11 : Among artists with available Spotify data, which artists have the highest Spotify popularity?

In [43]:
top_artists = (
    df.dropna(subset=['spotify_followers'])
      .sort_values('spotify_followers', ascending=False)
      [['name', 'spotify_followers']].head(10)
)

In [44]:
top_artists

,name,spotify_followers
10863,Taylor Swift,145489371.0
665,Ed Sheeran,122773292.0
48489,Billie Eilish,118692183.0
45810,The Weeknd,112957721.0
24358,Ariana Grande,107389514.0
10498,Eminem,104539036.0
16015,Drake,103039388.0
67344,Bad Bunny,102536542.0
16983,Justin Bieber,85008100.0
53017,BTS,81189603.0


In [45]:
alt.Chart(top_artists).mark_bar().encode(
    x = 'spotify_followers:Q',
    y= alt.Y('name:N', sort = alt.SortField(field = 'spotify_followers', order = 'descending')),
    color = 'name:N',
    tooltip = ['name', 'spotify_followers']
).properties(
    width = 500,
    height = 200
)

alt.Chart(...)

### Findings :
- The top artist with highest number of follower is "Taylor Swift" with 145489371 followers.

### Q12 : Among the 1.49% of artists with Spotify data, which genres have the highest average Spotify popularity and follower count?


In [46]:
Genre_avg_popularity = df.dropna(
    subset= ['spotify_followers']
    ).groupby('primary_genre_l1')['spotify_followers'].mean().reset_index(name = 'avg_follower').sort_values('avg_follower', ascending = False)

In [47]:
Genre_avg_popularity['avg_follower'] = Genre_avg_popularity['avg_follower'].round(0).astype('int64')

In [48]:
Genre_avg_popularity

,primary_genre_l1,avg_follower
7,Latin,20950612
5,Hip-Hop,9926661
9,Pop,9162600
12,Rock,6613398
3,Electronic,6595876
8,Metal,5375920
10,R&B,4250588
1,Classical,4071406
6,Jazz,3886038
2,Country,3099904


In [49]:
alt.Chart(Genre_avg_popularity).mark_bar().encode(
    x= alt.X('primary_genre_l1:N', sort = alt.SortField(field = 'avg_follower', order = 'descending'), axis = alt.Axis(labelAngle = 45)),
    y = 'avg_follower:Q',
    color = 'avg_follower:Q',
    tooltip = ['primary_genre_l1', 'avg_follower']
).properties(
    width = 500,
    height = 300
)

alt.Chart(...)

### Findings : 
- Among all the genres, "Latin" category has the highest average follower count with 20950612 average followers.
- Hip Hop comes second with 9926661 average followers which is less than half the average count of Latin.
- Pop stands at third place with 9162600 average followers, close to Hip Hop.